File: services/arca-fiscal-bridge/src/wsfe.ts
import type { ArcaConfig, ArcaResult, FiscalParameterSnapshot, FiscalParameterType, FiscalRequest, LoginTicket } from './types.js';
import { assertRemoteExecutionAllowed } from './config.js';
import { asArray, escapeXml, findFirst, parseTrustedSoap, xmlText } from './xml.js';
import { postSoap } from './transport.js';

const NS = 'http://ar.gov.afip.dif.FEV1/';

export const PARAMETER_OPERATIONS: Readonly<Record<FiscalParameterType, string>> = Object.freeze({
  document_types: 'FEParamGetTiposCbte',
  recipient_document_types: 'FEParamGetTiposDoc',
  vat_types: 'FEParamGetTiposIva',
  currencies: 'FEParamGetTiposMonedas',
  concepts: 'FEParamGetTiposConcepto',
  points_of_sale: 'FEParamGetPtosVenta',
});

export class WsfeClient {
  readonly #config: ArcaConfig;
  readonly #fetchImpl: typeof fetch;
  constructor(config: ArcaConfig, fetchImpl: typeof fetch = fetch) { this.#config = config; this.#fetchImpl = fetchImpl; }

  async dummy(): Promise<{ appServer: string; dbServer: string; authServer: string }> {
    const response = await this.#call('FEDummy', '<ar:FEDummy/>');
    const parsed = checkedSoap(response.body);
    return {
      appServer: xmlText(findFirst(parsed, 'AppServer')),
      dbServer: xmlText(findFirst(parsed, 'DbServer')),
      authServer: xmlText(findFirst(parsed, 'AuthServer')),
    };
  }

  async lastAuthorized(ticket: LoginTicket, pointOfSale: number, documentType: number): Promise<number> {
    assertPositiveInteger(pointOfSale, 'punto de venta');
    assertPositiveInteger(documentType, 'tipo de comprobante');
    const body = `<ar:FECompUltimoAutorizado>${authXml(ticket, this.#config.cuit)}<ar:PtoVta>${pointOfSale}</ar:PtoVta><ar:CbteTipo>${documentType}</ar:CbteTipo></ar:FECompUltimoAutorizado>`;
    const response = await this.#call('FECompUltimoAutorizado', body);
    const parsed = checkedSoap(response.body);
    const value = Number(xmlText(findFirst(parsed, 'CbteNro')));
    if (!Number.isSafeInteger(value) || value < 0) throw new Error('WSFEv1 devolvió último comprobante inválido.');
    return value;
  }

  async authorize(ticket: LoginTicket, request: FiscalRequest): Promise<ArcaResult> {
    validateFiscalRequest(request, this.#config.cuit);
    const response = await this.#call('FECAESolicitar', `<ar:FECAESolicitar>${authXml(ticket, request.cuit)}${caeRequestXml(request)}</ar:FECAESolicitar>`);
    return parseCaeResponse(response.body, response.requestHash, response.responseHash, request.documentNumber);
  }

  async consult(ticket: LoginTicket, request: Pick<FiscalRequest, 'cuit' | 'pointOfSale' | 'documentType' | 'documentNumber'>): Promise<ArcaResult | null> {
    validateCuit(request.cuit);
    const body = `<ar:FECompConsultar>${authXml(ticket, request.cuit)}<ar:FeCompConsReq><ar:CbteTipo>${request.documentType}</ar:CbteTipo><ar:CbteNro>${request.documentNumber}</ar:CbteNro><ar:PtoVta>${request.pointOfSale}</ar:PtoVta></ar:FeCompConsReq></ar:FECompConsultar>`;
    const response = await this.#call('FECompConsultar', body);
    const parsed = checkedSoap(response.body);
    const result = findFirst(parsed, 'ResultGet');
    if (!result || !xmlText(findFirst(result, 'CbteNro'))) return null;
    const cae = xmlText(findFirst(result, 'CodAutorizacion'));
    const totalAmount = numberOrUndefined(xmlText(findFirst(result, 'ImpTotal')));
    const pointOfSale = numberOrUndefined(xmlText(findFirst(result, 'PtoVta')));
    const documentType = numberOrUndefined(xmlText(findFirst(result, 'CbteTipo')));
    return {
      classification: /^\d{14}$/.test(cae) ? 'authorized' : 'service_error',
      documentNumber: Number(xmlText(findFirst(result, 'CbteNro'))),
      ...(totalAmount !== undefined ? { totalAmount } : {}),
      ...(pointOfSale !== undefined ? { pointOfSale } : {}),
      ...(documentType !== undefined ? { documentType } : {}),
      ...(cae ? { cae } : {}),
      ...(xmlText(findFirst(result, 'FchVto')) ? { caeExpiration: xmlText(findFirst(result, 'FchVto')) } : {}),
      ...(xmlText(findFirst(result, 'CbteFch')) ? { issueDate: xmlText(findFirst(result, 'CbteFch')) } : {}),
      observations: extractMessages(findFirst(result, 'Observaciones'), 'Obs'),
      errors: extractMessages(findFirst(parsed, 'Errors'), 'Err'),
      requestHash: response.requestHash,
      responseHash: response.responseHash,
    };
  }

  async getParameters(ticket: LoginTicket, parameterType: FiscalParameterType, now = new Date()): Promise<FiscalParameterSnapshot> {
    const operation = PARAMETER_OPERATIONS[parameterType];
    if (!operation) throw new Error('Tipo de parámetro WSFEv1 no permitido.');
    if (this.#config.environment === 'disabled') throw new Error('ARCA_DISABLED');
    const response = await this.#call(operation, `<ar:${operation}>${authXml(ticket, this.#config.cuit)}</ar:${operation}>`);
    const parsed = checkedSoap(response.body);
    const operationResult = findFirst(parsed, `${operation}Result`);
    const values = findFirst(operationResult, 'ResultGet');
    const errors = extractMessages(findFirst(operationResult, 'Errors'), 'Err');
    if (errors.length || values === undefined) {
      const error = new Error(errors.map(({ code, message }) => `${code}: ${message}`).join('; ') || 'WSFEv1 no devolvió la tabla solicitada.');
      Object.assign(error, { code: errors[0]?.code || 'MISSING_PARAMETER_RESULT', retryable: false });
      throw error;
    }
    return {
      environment: this.#config.environment,
      parameterType,
      operation,
      version: response.responseHash,
      synchronizedAt: now.toISOString(),
      values,
      requestHash: response.requestHash,
      responseHash: response.responseHash,
    };
  }

  async #call(operation: string, body: string) {
    assertRemoteExecutionAllowed(this.#config);
    return postSoap({
      endpoint: this.#config.endpoints.wsfe,
      action: `${NS}${operation}`,
      body: `<?xml version="1.0" encoding="utf-8"?><soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:ar="${NS}"><soapenv:Header/><soapenv:Body>${body}</soapenv:Body></soapenv:Envelope>`,
      fetchImpl: this.#fetchImpl,
    });
  }
}

export function validateFiscalRequest(request: FiscalRequest, configuredCuit: string): void {
  validateCuit(request.cuit);
  if (request.cuit !== configuredCuit) throw new Error('El CUIT del documento no coincide con el perfil del worker.');
  assertPositiveInteger(request.pointOfSale, 'punto de venta');
  assertPositiveInteger(request.documentType, 'tipo de comprobante');
  assertPositiveInteger(request.documentNumber, 'número');
  if (![1, 2, 3].includes(request.concept)) throw new Error('Concepto fiscal no soportado.');
  if (!/^\d{8}$/.test(request.issueDate)) throw new Error('Fecha fiscal inválida.');
  if (!/^[A-Z]{3}$/.test(request.currencyCode) || !(request.currencyRate > 0)) throw new Error('Moneda o cotización inválida.');
  const amounts = [request.totalAmount, request.netAmount, request.vatAmount, request.exemptAmount, request.nonTaxedAmount, request.otherTaxesAmount];
  if (amounts.some((value) => !Number.isFinite(value) || value < 0)) throw new Error('Importe fiscal inválido.');
  const components = request.netAmount + request.vatAmount + request.exemptAmount + request.nonTaxedAmount + request.otherTaxesAmount;
  if (Math.abs(cents(components) - cents(request.totalAmount)) > 1) throw new Error('El total fiscal no coincide con sus componentes.');
  if (request.vatItems.some((item) => !Number.isSafeInteger(item.id) || item.baseAmount < 0 || item.amount < 0)) throw new Error('Detalle IVA inválido.');
  if ((request.concept === 2 || request.concept === 3) && (!request.serviceFrom || !request.serviceTo || !request.paymentDueDate)) throw new Error('Servicios requieren período y vencimiento.');
  if (request.documentIntent === 'credit_note' && !request.associatedDocument) throw new Error('La nota de crédito requiere comprobante asociado.');
  if (request.associatedDocument) {
    assertPositiveInteger(request.associatedDocument.documentType, 'tipo asociado');
    assertPositiveInteger(request.associatedDocument.pointOfSale, 'punto de venta asociado');
    assertPositiveInteger(request.associatedDocument.documentNumber, 'número asociado');
    if (request.associatedDocument.cuit && !/^\d{11}$/.test(request.associatedDocument.cuit)) throw new Error('CUIT asociado inválido.');
    if (request.associatedDocument.issueDate && !/^\d{8}$/.test(request.associatedDocument.issueDate)) throw new Error('Fecha asociada inválida.');
  }
}

export function parseCaeResponse(xml: string, requestHash = '', responseHash = '', expectedNumber?: number): ArcaResult {
  const parsed = checkedSoap(xml);
  const detail = findFirst(parsed, 'FECAEDetResponse');
  const errors = extractMessages(findFirst(parsed, 'Errors'), 'Err');
  if (!detail) return { classification: 'service_error', observations: [], errors, requestHash, responseHash, errorCode: 'MISSING_DETAIL', errorMessage: 'WSFEv1 no devolvió detalle.' };
  const result = xmlText(findFirst(detail, 'Resultado'));
  const cae = xmlText(findFirst(detail, 'CAE'));
  const number = Number(xmlText(findFirst(detail, 'CbteDesde')) || expectedNumber);
  const observations = extractMessages(findFirst(detail, 'Observaciones'), 'Obs');
  if (result === 'A' && /^\d{14}$/.test(cae)) {
    return {
      classification: observations.length ? 'authorized_with_observations' : 'authorized',
      documentNumber: number,
      cae,
      caeExpiration: xmlText(findFirst(detail, 'CAEFchVto')),
      observations,
      errors,
      requestHash,
      responseHash,
    };
  }
  return {
    classification: result === 'R' ? 'rejected' : 'service_error',
    documentNumber: number,
    observations,
    errors,
    requestHash,
    responseHash,
    ...(!result ? { errorCode: 'MISSING_RESULT', errorMessage: 'WSFEv1 no informó Resultado.' } : {}),
  };
}

function caeRequestXml(request: FiscalRequest): string {
  const services = request.concept === 1 ? '' : `<ar:FchServDesde>${request.serviceFrom}</ar:FchServDesde><ar:FchServHasta>${request.serviceTo}</ar:FchServHasta><ar:FchVtoPago>${request.paymentDueDate}</ar:FchVtoPago>`;
  const vat = request.vatItems.length ? `<ar:Iva>${request.vatItems.map((item) => `<ar:AlicIva><ar:Id>${item.id}</ar:Id><ar:BaseImp>${money(item.baseAmount)}</ar:BaseImp><ar:Importe>${money(item.amount)}</ar:Importe></ar:AlicIva>`).join('')}</ar:Iva>` : '';
  const associated = request.associatedDocument ? `<ar:CbtesAsoc><ar:CbteAsoc><ar:Tipo>${request.associatedDocument.documentType}</ar:Tipo><ar:PtoVta>${request.associatedDocument.pointOfSale}</ar:PtoVta><ar:Nro>${request.associatedDocument.documentNumber}</ar:Nro>${request.associatedDocument.cuit ? `<ar:Cuit>${escapeXml(request.associatedDocument.cuit)}</ar:Cuit>` : ''}${request.associatedDocument.issueDate ? `<ar:CbteFch>${escapeXml(request.associatedDocument.issueDate)}</ar:CbteFch>` : ''}</ar:CbteAsoc></ar:CbtesAsoc>` : '';
  return `<ar:FeCAEReq><ar:FeCabReq><ar:CantReg>1</ar:CantReg><ar:PtoVta>${request.pointOfSale}</ar:PtoVta><ar:CbteTipo>${request.documentType}</ar:CbteTipo></ar:FeCabReq><ar:FeDetReq><ar:FECAEDetRequest><ar:Concepto>${request.concept}</ar:Concepto><ar:DocTipo>${request.recipientDocumentType}</ar:DocTipo><ar:DocNro>${escapeXml(request.recipientDocumentNumber)}</ar:DocNro><ar:CbteDesde>${request.documentNumber}</ar:CbteDesde><ar:CbteHasta>${request.documentNumber}</ar:CbteHasta><ar:CbteFch>${request.issueDate}</ar:CbteFch><ar:ImpTotal>${money(request.totalAmount)}</ar:ImpTotal><ar:ImpTotConc>${money(request.nonTaxedAmount)}</ar:ImpTotConc><ar:ImpNeto>${money(request.netAmount)}</ar:ImpNeto><ar:ImpOpEx>${money(request.exemptAmount)}</ar:ImpOpEx><ar:ImpIVA>${money(request.vatAmount)}</ar:ImpIVA><ar:ImpTrib>${money(request.otherTaxesAmount)}</ar:ImpTrib>${services}<ar:MonId>${escapeXml(request.currencyCode)}</ar:MonId><ar:MonCotiz>${request.currencyRate.toFixed(6)}</ar:MonCotiz>${associated}${vat}</ar:FECAEDetRequest></ar:FeDetReq></ar:FeCAEReq>`;
}

function checkedSoap(xml: string): unknown {
  const parsed = parseTrustedSoap(xml);
  const fault = findFirst(parsed, 'Fault');
  if (fault) {
    const message = xmlText(findFirst(fault, 'faultstring')) || 'ARCA devolvió SOAP Fault.';
    const error = new Error(message);
    Object.assign(error, { code: xmlText(findFirst(fault, 'faultcode')) || 'SOAP_FAULT', retryable: false });
    throw error;
  }
  return parsed;
}

function extractMessages(container: unknown, itemKey: string): Array<{ code: string; message: string }> {
  const items = asArray(findFirst(container, itemKey));
  return items.map((item) => ({ code: xmlText(findFirst(item, 'Code')), message: xmlText(findFirst(item, 'Msg')) })).filter((item) => item.code || item.message);
}

function authXml(ticket: LoginTicket, cuit: string): string { return `<ar:Auth><ar:Token>${escapeXml(ticket.token)}</ar:Token><ar:Sign>${escapeXml(ticket.sign)}</ar:Sign><ar:Cuit>${escapeXml(cuit)}</ar:Cuit></ar:Auth>`; }
function validateCuit(value: string): void { if (!/^\d{11}$/.test(value)) throw new Error('CUIT inválido.'); }
function assertPositiveInteger(value: number, label: string): void { if (!Number.isSafeInteger(value) || value < 1) throw new Error(`${label} inválido.`); }
function cents(value: number): number { return Math.round((value + Number.EPSILON) * 100); }
function money(value: number): string { return (Math.round((value + Number.EPSILON) * 100) / 100).toFixed(2); }
function numberOrUndefined(value: string): number | undefined { const parsed = Number(value); return Number.isFinite(parsed) ? parsed : undefined; }

File: agents/windows-local-agent/src/Taba.LocalAgent.Core/Fiscal/Wsfe.cs
using System.Globalization;
using System.Text;
using System.Xml;
using System.Xml.Linq;

namespace Taba.LocalAgent.Core.Fiscal;

/// <summary>Una alícuota de IVA del comprobante (Id según <c>FEParamGetTiposIva</c>).</summary>
public sealed record VatLine(int Id, decimal TaxableBase, decimal Amount);

/// <summary>Comprobante asociado (una nota de crédito apunta a su factura).</summary>
public sealed record AssociatedVoucher(int VoucherType, int PointOfSale, long Number, string IssuerCuit, DateOnly IssueDate);

/// <summary>
/// Pedido de CAE para UN comprobante. Todos los códigos (tipo, documento,
/// condición frente al IVA, alícuotas) vienen de la política fiscal aprobada y
/// de los parámetros oficiales sincronizados: este modelo no infiere ninguno.
/// </summary>
public sealed record WsfeInvoiceRequest
{
    public required string IssuerCuit { get; init; }

    public required int PointOfSale { get; init; }

    public required int VoucherType { get; init; }

    /// <summary>1 Productos, 2 Servicios, 3 Productos y Servicios.</summary>
    public required int Concept { get; init; }

    public required int RecipientDocType { get; init; }

    public required long RecipientDocNumber { get; init; }

    /// <summary>Número reservado localmente antes de llamar a ARCA.</summary>
    public required long VoucherNumber { get; init; }

    public required DateOnly IssueDate { get; init; }

    public required decimal Total { get; init; }

    public decimal NonTaxed { get; init; }

    public required decimal Net { get; init; }

    public decimal Exempt { get; init; }

    public decimal OtherTaxes { get; init; }

    public required decimal Vat { get; init; }

    public string Currency { get; init; } = "PES";

    public decimal CurrencyRate { get; init; } = 1m;

    /// <summary>
    /// Condición frente al IVA del receptor (manual 4.0, RG 5616). Se valida
    /// contra <c>FEParamGetCondicionIvaReceptor</c>; ARCA rechaza con 10246 cuando falta.
    /// </summary>
    public required int ReceiverVatConditionId { get; init; }

    public IReadOnlyList<VatLine> VatLines { get; init; } = [];

    public AssociatedVoucher? Associated { get; init; }

    public DateOnly? ServiceFrom { get; init; }

    public DateOnly? ServiceTo { get; init; }

    public DateOnly? PaymentDue { get; init; }
}

/// <summary>Arma los sobres SOAP de WSFEv1 con XmlWriter: ningún dato se concatena como texto.</summary>
public static class WsfeSoap
{
    public const string Namespace = "http://ar.gov.afip.dif.FEV1/";
    private const string SoapNamespace = "http://schemas.xmlsoap.org/soap/envelope/";

    public static string SoapAction(string operation) => Namespace + operation;

    public static void Validate(WsfeInvoiceRequest request)
    {
        ArgumentNullException.ThrowIfNull(request);
        var errors = new List<string>();
        if (request.IssuerCuit.Length != 11 || !request.IssuerCuit.All(char.IsAsciiDigit))
        {
            errors.Add("ISSUER_CUIT_INVALID");
        }

        if (request.PointOfSale is < 1 or > 99999)
        {
            errors.Add("POINT_OF_SALE_INVALID");
        }

        if (request.VoucherType <= 0)
        {
            errors.Add("VOUCHER_TYPE_REQUIRED");
        }

        if (request.Concept is < 1 or > 3)
        {
            errors.Add("CONCEPT_INVALID");
        }

        if (request.VoucherNumber is < 1 or > 99_999_999)
        {
            errors.Add("VOUCHER_NUMBER_INVALID");
        }

        if (request.ReceiverVatConditionId <= 0)
        {
            errors.Add("RECEIVER_VAT_CONDITION_REQUIRED");
        }

        if (request.Concept is 2 or 3 && (request.ServiceFrom is null || request.ServiceTo is null || request.PaymentDue is null))
        {
            errors.Add("SERVICE_DATES_REQUIRED");
        }

        // ImpTotal = ImpTotConc + ImpNeto + ImpOpEx + ImpTrib + ImpIVA (regla del manual).
        var sum = request.NonTaxed + request.Net + request.Exempt + request.OtherTaxes + request.Vat;
        if (decimal.Round(sum, 2, MidpointRounding.ToEven) != decimal.Round(request.Total, 2, MidpointRounding.ToEven))
        {
            errors.Add("TOTAL_MISMATCH");
        }

        var vatSum = request.VatLines.Sum(v => v.Amount);
        if (request.VatLines.Count > 0 && decimal.Round(vatSum, 2, MidpointRounding.ToEven) != decimal.Round(request.Vat, 2, MidpointRounding.ToEven))
        {
            errors.Add("VAT_LINES_MISMATCH");
        }

        if (errors.Count > 0)
        {
            throw new ArgumentException(string.Join(",", errors), nameof(request));
        }
    }

    public static string FeCaeSolicitar(AccessTicket ticket, WsfeInvoiceRequest request)
    {
        Validate(request);
        return Envelope("FECAESolicitar", ticket, request.IssuerCuit, w =>
        {
            w.WriteStartElement("ar", "FeCAEReq", Namespace);
            w.WriteStartElement("ar", "FeCabReq", Namespace);
            Element(w, "CantReg", 1);
            Element(w, "PtoVta", request.PointOfSale);
            Element(w, "CbteTipo", request.VoucherType);
            w.WriteEndElement();
            w.WriteStartElement("ar", "FeDetReq", Namespace);
            w.WriteStartElement("ar", "FECAEDetRequest", Namespace);
            Element(w, "Concepto", request.Concept);
            Element(w, "DocTipo", request.RecipientDocType);
            Element(w, "DocNro", request.RecipientDocNumber);
            Element(w, "CbteDesde", request.VoucherNumber);
            Element(w, "CbteHasta", request.VoucherNumber);
            Element(w, "CbteFch", Date(request.IssueDate));
            Element(w, "ImpTotal", Amount(request.Total));
            Element(w, "ImpTotConc", Amount(request.NonTaxed));
            Element(w, "ImpNeto", Amount(request.Net));
            Element(w, "ImpOpEx", Amount(request.Exempt));
            Element(w, "ImpTrib", Amount(request.OtherTaxes));
            Element(w, "ImpIVA", Amount(request.Vat));
            if (request.Concept is 2 or 3)
            {
                Element(w, "FchServDesde", Date(request.ServiceFrom!.Value));
                Element(w, "FchServHasta", Date(request.ServiceTo!.Value));
                Element(w, "FchVtoPago", Date(request.PaymentDue!.Value));
            }

            Element(w, "MonId", request.Currency);
            Element(w, "MonCotiz", request.CurrencyRate.ToString("0.######", CultureInfo.InvariantCulture));
            // En el orden del WSDL: CanMisMonExt (no aplica a PES) y la condición del receptor.
            Element(w, "CondicionIVAReceptorId", request.ReceiverVatConditionId);
            if (request.Associated is { } associated)
            {
                w.WriteStartElement("ar", "CbtesAsoc", Namespace);
                w.WriteStartElement("ar", "CbteAsoc", Namespace);
                Element(w, "Tipo", associated.VoucherType);
                Element(w, "PtoVta", associated.PointOfSale);
                Element(w, "Nro", associated.Number);
                Element(w, "Cuit", associated.IssuerCuit);
                Element(w, "CbteFch", Date(associated.IssueDate));
                w.WriteEndElement();
                w.WriteEndElement();
            }

            if (request.VatLines.Count > 0)
            {
                w.WriteStartElement("ar", "Iva", Namespace);
                foreach (var vat in request.VatLines)
                {
                    w.WriteStartElement("ar", "AlicIva", Namespace);
                    Element(w, "Id", vat.Id);
                    Element(w, "BaseImp", Amount(vat.TaxableBase));
                    Element(w, "Importe", Amount(vat.Amount));
                    w.WriteEndElement();
                }

                w.WriteEndElement();
            }

            w.WriteEndElement();
            w.WriteEndElement();
            w.WriteEndElement();
        });
    }

    public static string FeCompConsultar(AccessTicket ticket, string cuit, int voucherType, int pointOfSale, long number) =>
        Envelope("FECompConsultar", ticket, cuit, w =>
        {
            w.WriteStartElement("ar", "FeCompConsReq", Namespace);
            Element(w, "CbteTipo", voucherType);
            Element(w, "CbteNro", number);
            Element(w, "PtoVta", pointOfSale);
            w.WriteEndElement();
        });

    public static string FeCompUltimoAutorizado(AccessTicket ticket, string cuit, int pointOfSale, int voucherType) =>
        Envelope("FECompUltimoAutorizado", ticket, cuit, w =>
        {
            Element(w, "PtoVta", pointOfSale);
            Element(w, "CbteTipo", voucherType);
        });

    private static string Envelope(string operation, AccessTicket ticket, string cuit, Action<XmlWriter> body)
    {
        ArgumentNullException.ThrowIfNull(ticket);
        var builder = new StringBuilder();
        var settings = new XmlWriterSettings { OmitXmlDeclaration = true, Encoding = Encoding.UTF8 };
        using (var w = XmlWriter.Create(builder, settings))
        {
            w.WriteStartElement("soapenv", "Envelope", SoapNamespace);
            w.WriteAttributeString("xmlns", "ar", null, Namespace);
            w.WriteElementString("soapenv", "Header", SoapNamespace, string.Empty);
            w.WriteStartElement("soapenv", "Body", SoapNamespace);
            w.WriteStartElement("ar", operation, Namespace);
            w.WriteStartElement("ar", "Auth", Namespace);
            Element(w, "Token", ticket.Token);
            Element(w, "Sign", ticket.Sign);
            Element(w, "Cuit", cuit);
            w.WriteEndElement();
            body(w);
            w.WriteEndElement();
            w.WriteEndElement();
            w.WriteEndElement();
        }

        return builder.ToString();
    }

    private static void Element(XmlWriter writer, string name, object value) =>
        writer.WriteElementString("ar", name, Namespace, Convert.ToString(value, CultureInfo.InvariantCulture));

    private static string Date(DateOnly date) => date.ToString("yyyyMMdd", CultureInfo.InvariantCulture);

    // Redondeo half-even a dos decimales, como declara el manual.
    private static string Amount(decimal value) =>
        decimal.Round(value, 2, MidpointRounding.ToEven).ToString("0.00", CultureInfo.InvariantCulture);
}

public enum CaeOutcome
{
    Approved,
    Rejected,
}

public sealed record ArcaMessage(int Code, string Message);

/// <summary>El resultado de FECAESolicitar para un comprobante.</summary>
public sealed record CaeResult(
    CaeOutcome Outcome,
    long VoucherNumber,
    string? Cae,
    DateOnly? CaeExpiration,
    IReadOnlyList<ArcaMessage> Observations,
    IReadOnlyList<ArcaMessage> Errors);

/// <summary>Un comprobante tal como lo devuelve FECompConsultar.</summary>
public sealed record ConsultedVoucher(
    int VoucherType,
    int PointOfSale,
    long Number,
    DateOnly IssueDate,
    decimal Total,
    int RecipientDocType,
    long RecipientDocNumber,
    string Result,
    string? Cae,
    DateOnly? CaeExpiration);

/// <summary>Lectura de las respuestas de WSFEv1. Una aprobación sin CAE válido es un error de protocolo.</summary>
public static class WsfeResponses
{
    private static readonly XNamespace Ns = WsfeSoap.Namespace;

    public const int NotFoundCode = 602;

    public static CaeResult ParseCaeResponse(string soapXml)
    {
        var result = Result(soapXml, "FECAESolicitarResult");
        var errors = Messages(result.Element(Ns + "Errors"), "Err");
        var detail = result.Element(Ns + "FeDetResp")?.Element(Ns + "FEDetResponse");
        if (detail is null)
        {
            if (errors.Count == 0)
            {
                throw new FormatException("WSFE_RESPONSE_WITHOUT_DETAIL");
            }

            return new CaeResult(CaeOutcome.Rejected, 0, null, null, [], errors);
        }

        var observations = Messages(detail.Element(Ns + "Obs"), "Observaciones");
        var number = long.Parse(detail.Element(Ns + "CbteDesde")?.Value ?? "0", CultureInfo.InvariantCulture);
        var outcome = detail.Element(Ns + "Resultado")?.Value;
        if (outcome == "A")
        {
            var cae = detail.Element(Ns + "CAE")?.Value?.Trim();
            if (cae is null || cae.Length != 14 || !cae.All(char.IsAsciiDigit))
            {
                throw new FormatException("WSFE_APPROVED_WITHOUT_VALID_CAE");
            }

            return new CaeResult(CaeOutcome.Approved, number, cae, ParseDate(detail.Element(Ns + "CAEFchVto")?.Value), observations, errors);
        }

        return new CaeResult(CaeOutcome.Rejected, number, null, null, observations, errors);
    }

    /// <summary>El comprobante consultado, o null si ARCA contesta 602 (no existe).</summary>
    public static ConsultedVoucher? ParseConsultResponse(string soapXml)
    {
        var result = Result(soapXml, "FECompConsultarResult");
        var errors = Messages(result.Element(Ns + "Errors"), "Err");
        var get = result.Element(Ns + "ResultGet");
        if (get is null)
        {
            if (errors.Any(e => e.Code == NotFoundCode))
            {
                return null;
            }

            throw new FormatException(errors.Count > 0 ? $"WSFE_ERROR_{errors[0].Code}" : "WSFE_CONSULT_WITHOUT_RESULT");
        }

        var number = long.Parse(get.Element(Ns + "CbteDesde")?.Value ?? "0", CultureInfo.InvariantCulture);
        var cae = get.Element(Ns + "CodAutorizacion")?.Value?.Trim();
        return new ConsultedVoucher(
            int.Parse(get.Element(Ns + "CbteTipo")?.Value ?? "0", CultureInfo.InvariantCulture),
            int.Parse(get.Element(Ns + "PtoVta")?.Value ?? "0", CultureInfo.InvariantCulture),
            number,
            ParseDate(get.Element(Ns + "CbteFch")?.Value) ?? throw new FormatException("WSFE_CONSULT_WITHOUT_DATE"),
            decimal.Parse(get.Element(Ns + "ImpTotal")?.Value ?? "0", CultureInfo.InvariantCulture),
            int.Parse(get.Element(Ns + "DocTipo")?.Value ?? "0", CultureInfo.InvariantCulture),
            long.Parse(get.Element(Ns + "DocNro")?.Value ?? "0", CultureInfo.InvariantCulture),
            get.Element(Ns + "Resultado")?.Value ?? string.Empty,
            string.IsNullOrEmpty(cae) ? null : cae,
            ParseDate(get.Element(Ns + "FchVto")?.Value));
    }

    public static long ParseLastAuthorized(string soapXml)
    {
        var result = Result(soapXml, "FECompUltimoAutorizadoResult");
        var errors = Messages(result.Element(Ns + "Errors"), "Err");
        var number = result.Element(Ns + "CbteNro")?.Value;
        if (number is null)
        {
            throw new FormatException(errors.Count > 0 ? $"WSFE_ERROR_{errors[0].Code}" : "WSFE_LAST_WITHOUT_NUMBER");
        }

        return long.Parse(number, CultureInfo.InvariantCulture);
    }

    private static XElement Result(string soapXml, string resultName) =>
        SafeXml.Parse(soapXml).Descendants(Ns + resultName).FirstOrDefault()
        ?? throw new FormatException($"WSFE_MISSING_{resultName}");

    private static List<ArcaMessage> Messages(XElement? container, string item) =>
        container?.Elements(Ns + item)
            .Select(e => new ArcaMessage(
                int.Parse(e.Element(Ns + "Code")?.Value ?? "0", CultureInfo.InvariantCulture),
                e.Element(Ns + "Msg")?.Value ?? string.Empty))
            .ToList() ?? [];

    private static DateOnly? ParseDate(string? value) =>
        DateOnly.TryParseExact(value?.Trim(), "yyyyMMdd", CultureInfo.InvariantCulture, DateTimeStyles.None, out var date) ? date : null;
}
